src/mysql_conn.erl

```diff
@@ -91,54 +92,109 @@ init(Opts) ->
         N when N > 0 -> N
     end,
 
+    State0 = #state{
+        tcp_opts = TcpOpts,
+        ssl_opts = SSLOpts,
+        host = Host, port = Port,
+        user = User, password = Password,
+        database = Database,
+        queries = Queries, prepares = Prepares,
+        log_warnings = LogWarn,
+        connect_timeout = ConnectTimeout,
+        ping_timeout = PingTimeout,
+        query_timeout = QueryTimeout,
+        query_cache_time = QueryCacheTime,
+        cap_found_rows = (SetFoundRows =:= true)
+    },
+
+    case proplists:get_value(connect_mode, Opts, synchronous) of
+        synchronous ->
+            case do_connect(State0) of
+                {ok, State1} ->
+                    {ok, State1};
+                {error, Reason} ->
+                    {stop, Reason}
+            end;
+        asynchronous ->
+            gen_server:cast(self(), connect),
+            {ok, State0};
+        lazy ->
+            {ok, State0}
+    end.
+
+do_connect(#state{connect_timeout = ConnectTimeout} = State) ->
+    Self = self(),
+    Pid = spawn_link(
+        fun () ->
+            case do_connect1(State) of
+                {ok, #state{sockmod = SockMod, socket = Socket} = State1} ->
+                    SockMod:controlling_process(Socket, Self),
+                    Self ! {self(), {ok, State1}};
+                {error, _} = E ->
+                    Self ! {self(), E}
+            end
+        end
+    ),
+    receive
+        {Pid, {ok, State1}} ->
+            do_post_connect(State1);
+        {Pid, {error, _} = E} ->
+            E
+    after ConnectTimeout ->
+        unlink(Pid),
+        exit(Pid, kill),
+        {error, timeout}
+    end.
+
+do_connect1(#state{tcp_opts = TcpOpts, host = Host, port = Port} = State) ->
```

Comment:
The names do_connect, do_connect1, do_connect2 could be better, e.g. by including words like handshake, depending on what they do. Comments (or edoc) can also be used to describe them if the names are not clear enough.

Reply:
In fact, I had it like this before, forgot why I switched over to 1, 2, ... =^^=

Reply:
Ah, but please think about the names and possibly split them further.

Any names like connection, authenticate, do_handshake, pre_connect, etc. are better IMO. With numbers you can't split a function in two without changing all the numbers. And "connecting" is actually just the tcp part. All the rest is something else, if you think about it.

Reply:
> And "connecting" is actually just the tcp part. All the rest is something else, if you think about it.

I know, but what would you call the overall process of connecting, authentication, ssl upgrade, etc? The one function that wraps all that, I mean. Nothing comes to mind for me =^^=

Reply:
Hm... ok, I will think about it. Perhaps I can come up with something.

Reply:
When the functions are chained using tail calls, it's hard to name them and dependency-wise they all depend on each other. Thus, I suggest not chaining them using tail calls.

Instead call them one by one from `do_connect` (which I think can simply be called `connect`). Rename `do_connect1` => `connect_tcp` and let it return the updated state with the socket set. Then rename `do_connect2` => `do_handshake` and don't call `do_post_connect`. `do_post_connect` can also be called directly from `connect` (but a case is needed for reporting handshake errors properly).

This puts some more code in the already deeply indented argument to spawn_link. This can be moved to it's own function, which can take the name `do_connect` (which in my world indicates that it's a helper to the `connect` function).

```
connect     do_connect
   |           |
   |---spawn-->|     
   |           |<-------> connect_tcp
   |           |<-------> do_handshake
   |           |<-------> do_post_connect
   |<-receive--|
```
instead of
```
do_connect   fun()
    |          |
    |--spawn-->|
    |          |--> do_connect1 -> do_connect2 -> do_post_connect
    |          |<----(return)--------------------------|
    |<-receive-|
```

Reply:
Hm, I made some changes just before receiving this comment. I renamed `do_connect` to `connect`, `do_connect1` to `connect_socket`, `do_connect2` to `handshake`, and `do_post_connect` to `post_connect`.

Other than you depict, `post_connect` is not done in the spawned process but after the receive, and is thus not affected by the connect timeout. This is for two reasons: For one, I'd say it is not part of the actual connection procedure, but afterwards, and thus should not be affected by the connect timeout; for another, this has been the behavior before the on-connect prepares and queries were moved into the connection process (IIRC, by my last PR before this one), putting the on-connect stuff under the connect timeout was an accident.

Reply:
Really good! I agree post_connect should be after the receive.